### PR TITLE
intermediate specs and parsing

### PR DESCRIPTION
**Pull request by jeroenrinzema**
Intermediate specs (HCL) and parsing of specs to the spec data structure.

### CODE REVIEW

File: specs/template.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+const (
+	// TemplateOpen tag
+	TemplateOpen = "{{"
+	// TemplateClose tag
+	TemplateClose = "}}"
+	// ReferenceDelimiter represents the value resource reference delimiter
+	ReferenceDelimiter = ":"
+	// PathDelimiter represents the path reference delimiter
+	PathDelimiter = "."
+
```
Comment (singhashmeet) [inline specs/template.go]: Shouldn’t we add UserDefinedFunctionOpen = "(" and UserDefinedFunctionClose = ")" over here

**Comment by jeroenrinzema** [comment]
I have included support for custom-defined functions, there is still lacking support for constant types such as `'string'`, `1` and `true` but this could be added in a later version.

File: specs/template.go (hunk 1 above)
Comment (jeroenrinzema) [reply specs/template.go]: The custom defined functions are pattern matched using regex, the pattern is stored in `FunctionPattern`.

File: specs/intermediate/intermediate.go [hunk 2]
```diff
@@ -64,11 +65,12 @@ type RepeatedParameterMap struct {
 
 // Call intermediate specification
 type Call struct {
-	Name     string        `hcl:"name,label"`
-	Endpoint string        `hcl:"endpoint,label"`
-	Type     string        `hcl:"type,optional"`
-	Request  *ParameterMap `hcl:"request,block"`
-	Rollback *RollbackCall `hcl:"rollback,block"`
+	DependsOn []string      `hcl:"depends_on,optional"`
```
Comment (velp) [inline specs/intermediate/intermediate.go]: Do we able to run several calls in parallel mode if they do not have cross-dependencies?

Comment (jeroenrinzema) [reply specs/intermediate/intermediate.go]: I got a check for cycle dependencies available inside a future PR. For now, we only need to worry about the possibility of defining dependencies.

**Comment by singhashmeet** [comment]
Approved based on the meeting we had

**Pull request merged**